### ISSUE TITLE
Proposed rule: no-await-in-loop

### ISSUE THREAD

**Issue by nmote** [eslint/eslint#4166]
**When does this rule warn?**

This rule warns when a developer uses `await` inside of a loop. Doing so means that each loop iteration will run in serial, when often the desired behavior is to dispatch a bunch of asynchronous operations and then await on the result as a whole, using `Promise.all`.

Occasionally, awaiting inside a loop actually is desirable. In cases like those, 

Here is an example:

``` js
async foo() {
  for (const bar of baz) {
    await(bar);
  }
}
```

Here is my proposed warning:

'Avoid using await inside a loop. Consider refactoring to use Promise.all. If you are sure you want to do this, add `// eslint-disable-line no-await-in-loop` at the end of this line.'

**Is this rule preventing an error or is it stylistic?**

This rule prevents an error. Needlessly awaiting inside of a loop can have serious performance implications.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**

Using `await` inside of a loop is a bad pattern no matter what codebase you are operating in. There are occasional places where it is desirable but it is almost always a sign that you have done something wrong.

My only concern is that this rule will not operate with the `eslint` frontend -- to my knowledge, it has not implemented async/await parsing yet. Therefore it could be confusing to have a rule built-in to `eslint` that requires its users to use `babel-eslint`.

**Are you willing to create the rule yourself?**

Yes. I have already written the [first version](https://github.com/facebook/nuclide/blob/4c83f145469ef52b384839ce03e8913ba76ae9f0/scripts/lib/eslint/eslint_rules/no-await-in-loop.js).

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by hzoo** [comment]
You can add it to https://github.com/babel/eslint-plugin-babel if not here

**Comment by ilyavolodin** [comment]
ESLint doesn't natively supports experimental features. As such, we can't create core rules for features beyond ES2015. Link that @hzoo posted is probably the best bet for such a rule.

**Comment by zertosh** [comment]
Now that async/await is supported in eslint, should this issue be re-opened and re-considered?

**Comment by platinumazure** [comment]
Yes, we should reconsider this. Reopening.

**Comment by not-an-aardvark** [comment]
Nit: I think `no-loop-await` might be a better name, since it parallels [`no-loop-func`](http://eslint.org/docs/rules/no-loop-func).

**Comment by nmote** [comment]
I contributed it [here](https://github.com/babel/eslint-plugin-babel/blob/master/rules/no-await-in-loop.js) and I'd be happy to put up a PR to get it in ESLint. The proposed name change is fine with me.

**Comment by platinumazure** [comment]
Personally I like `no-await-in-loop` better (it's longer but clearer). And if I could change `no-loop-func` to `no-func-in-loop`, I would. But that's a minor concern. :smile:

Anyway, I'm endorsing this (via :+1: to the original issue comment). With two more and a champion, we can accept this.

**Comment by kaicataldo** [comment]
This is definitely a useful thing to guard against. I'm curious, though, if we have any other rules that require the user to manually ignore legitimate use cases because the rule can't sniff these out on their own. Not entirely convinced this meets the bar for core or should be recommended for a plugin.

**Comment by kaicataldo** [comment]
Ah, right, there's `default-case`. I'm still a little wary of forcing users to use disable comments, but I'm not sure there's a better way.

**Comment by ljharb** [comment]
@kaicataldo there's also http://eslint.org/docs/rules/no-empty for empty `catch` statements.

**Comment by nmote** [comment]
I was going to finally put up the PR today but I'll wait for consensus on whether it belongs in core.

**Comment by kaicataldo** [comment]
Sorry for holding this up. I'm still not a huge fan of relying on the built in `eslint-disable` comments (rather than a rule-specific thing), but given that's my only concern, I support this enhancement. We can always add a rule-specific way to disable if it feels necessary in the future.

**Comment by nmote** [comment]
I'm open to suggestions on how to implement the opt-out in a more convenient way for users. Personally I think that using the same syntax for disabling rules across the board is a simpler experience but my mind is open.

Since there's no consensus on the name I'll leave it as `no-await-in-loop`. I agree with @platinumazure that it's clearer. Again, happy to entertain discussion if anybody feels strongly.

I'll put up a PR on Monday.

**Comment by kaicataldo** [comment]
I'll champion this.

**Comment by kaicataldo** [comment]
Whoops, miscounted. @eslint/eslint-team We need one more :+1:

**Comment by vitorbal** [comment]
I will :+1: this

**Comment by kaicataldo** [comment]
@nmote Thanks! This issue is now accepted - let me know if you have any issues.

**Comment by nmote** [comment]
It's taking a bit longer to check all the boxes than I expected. I'll finish this up tomorrow.
